### DISCUSSION TITLE
Quasar integration with vitest

### DISCUSSION THREAD

**Discussion by DanielParra05** [quasarframework/quasar#12810] [General - Components / Directives / etc] [answered]
Hello, 
I have been trying to integrate the unit testing framework [**vitest** ](https://vitest.dev/) with a project that implements quasar, but I have not succeed doing so. The main problem that I am facing when testing is that quasar components are not rendering in HTML elements, so when I try to set a text on an <q-input> element **vitest** does not identify it as a HTML element and I get the next error: 

![image](https://user-images.githubusercontent.com/37990069/158425771-b6806a00-9053-4244-8b92-61f6a014b9da.png)

I tried a console.log(wrapper.get('[label="Email"]').html()) and I got the follow: 
![image](https://user-images.githubusercontent.com/37990069/158426676-1e1b7e65-725c-474b-9fe0-e84eb251ec2c.png)

As you can see the element is not being "transformed" to an HTML tag. Is it possible an integration of quasar with vitest? If it is, could you please let me know how it should be ?

TIA

**Comment by yusufkandemir** [comment]
You need to install the Quasar plugin and any other package/plugin/library that you are using. `mount` can't automatically guess and include them. Here is how it's being handled in Quasar Testing Jest AE:

https://github.com/quasarframework/quasar-testing/tree/6216c3390164b41bf19ccf3658014558f216841f/packages/unit-jest#installquasarpluginoptions
https://github.com/quasarframework/quasar-testing/blob/6216c3390164b41bf19ccf3658014558f216841f/packages/unit-jest/src/helpers/install-quasar-plugin.ts

Here is the [Plugins docs page for Vue Test Utils](https://test-utils.vuejs.org/guide/extending-vtu/plugins.html).

**Reply by DanielParra05** [reply]
Thank you for your response. I made some adjustments similar to the ones you show me that were implemented in Quasar Testing Jest AE, but using Vitest insetad of Jest. I created a separate function that injects the plugin into the test-utils config and I call this function from the test file. 
Implementation of installQuasarPlugin: 

```
import { afterAll } from 'vitest';
import { config } from '@vue/test-utils';
import { cloneDeep } from 'lodash-es';
import { Quasar, QuasarPluginOptions } from 'quasar';

export function installQuasarPlugin(options?: Partial<QuasarPluginOptions>) {
    const globalConfigBackup = cloneDeep(config.global);
    config.global.plugins.unshift([Quasar, options]);

    afterAll(() => {
        config.global = globalConfigBackup;
    });
}
```

My test file: 
```
import { test, expect, describe } from 'vitest'
import { mount } from '@vue/test-utils'
import App from '../App.vue'
import { installQuasarPlugin } from '../install-quasar-plugin';

installQuasarPlugin();
const wrapper = mount(App);

describe('tests for app', () => {
  test('mount component', async () => {
    console.log("Buton HTML--->  " + wrapper.get('#testButton').html());

    expect(wrapper).toBeTruthy();
  })
})
```

I was expecting that now the button will print as HTML tag without problem since the plug in was inserted. But now the component is not mounting and I get the following error: 
```
 FAIL  src/tests/testvite.spec.ts [ src/tests/testvite.spec.ts ]
TypeError: Cannot convert undefined or null to object
 ❯ installQuasar node_modules/quasar/dist/quasar.cjs.prod.js:6:15243
 ❯ Object.install node_modules/quasar/dist/quasar.cjs.prod.js:6:477236
 ❯ Object.use node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:3790:28
 ❯ Proxy.mount node_modules/@vue/test-utils/dist/vue-test-utils.cjs.js:7860:25
 ❯ src/tests/testvite.spec.ts:7:16
      5| 
      6| installQuasarPlugin();
      7| const wrapper = mount(App);
       |                ^
      8| 
      9| describe('tests for app', () => {
 ❯ async collectTests node_modules/vitest/dist/entry.js:1600:7
 ❯ async startTests node_modules/vitest/dist/entry.js:1870:17
```
Do you have any idea what I might be doing wrong?

**Comment by DanielParra05** [comment, marked as answer]
For anyone facing the same issue, I just made it work with @yusufkandemir recommendation but with a different approach than the one I showed before. I'll let my code below:
vite.config.ts:
```
import { defineConfig } from 'vitest/config'
import vue from '@vitejs/plugin-vue'
import { quasar, transformAssetUrls } from '@quasar/vite-plugin'

// https://vitejs.dev/config/
export default defineConfig({
  test: {
    environment: 'jsdom'
  },
  plugins: [
    vue({
      template: { transformAssetUrls }
    }),
    quasar({
      sassVariables: 'src/quasar-variables.sass'
    })
  ],
})
```

Implementation of quasar plugin in test:
```
import { test, expect, describe } from 'vitest'
import { mount } from '@vue/test-utils'
import { Quasar } from 'quasar'

import HelloWorld from "../components/HelloWorld.vue"

const wrapperFactory = () => mount(HelloWorld, {
    global: {
        plugins: [Quasar]
    },
})

test('mount component', () => {
    expect(HelloWorld).toBeTruthy();
    const wrapper = wrapperFactory();

    console.log(wrapper.html());
})
```
Now, when it prints the html from the wrapper, Quasar elements are being shown as simple HTML elements. 💯

**Reply by chiaraani** [reply]
How can I add options to Quasar plugin installation?
```js
app.use(Quasar, {  plugins: { LocalStorage } })
```
How can I test that?

I tried this but `TypeError: Cannot read properties of undefined (reading 'getItem')`.
```js
import { render } from '@testing-library/vue'

const wrapper = render(DrawerBtn, {
    global: { plugins: [Quasar, { plugins: { LocalStorage } }] }
  })
```
I also tried this but it does not work to me. It throws same error.
```js
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
installQuasarPlugin({ plugins: { LocalStorage } })
```
I'm using `@quasar/vite-plugin`. Please, help.

**Comment by prabhatsharma** [comment]
It takes me forward but does not solve the whole problem. I tried below.

```js
import { mount } from "@vue/test-utils";
import About from "../About.vue";

import { describe, expect, it } from "vitest";
import { Quasar } from 'quasar'

import store from "../../store";

describe("About", () => {
  it("renders correctly", () => {
    const wrapper = () => mount(About, {
      global: {
        plugins: [Quasar, store]
      },
    })

    console.log(wrapper().html());
  });
});
```

I am getting this error at console.log(wrapper().html()); :

```shell
FAIL  src/views/__tests__/About.spec.ts > About > renders correctly
TypeError: Cannot read property 'header' of undefined
 ❯ ReactiveEffect.fn node_modules/quasar/src/components/page/QPage.js:25:20
     23|     const style = computed(() => {
     24|       const offset
     25|         = ($layout.header.space === true ? $layout.header.size : 0)
       |                    ^
     26|         + ($layout.footer.space === true ? $layout.footer.size : 0)
     27| 
 ❯ ReactiveEffect.run node_modules/@vue/reactivity/dist/reactivity.cjs.js:189:25
 ❯ ComputedRefImpl.get value [as value] node_modules/@vue/reactivity/dist/reactivity.cjs.js:1115:39
 ❯ Proxy.<anonymous> node_modules/quasar/src/components/page/QPage.js:53:20
 ❯ renderComponentRoot node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:889:44
 ❯ ReactiveEffect.componentUpdateFn [as fn] node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:4990:57
 ❯ ReactiveEffect.run node_modules/@vue/reactivity/dist/reactivity.cjs.js:189:25
 ❯ setupRenderEffect node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:5116:9
 ❯ mountComponent node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:4899:9
 ❯ processComponent node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:4857:17
```

Any ideas on what could be going on?
